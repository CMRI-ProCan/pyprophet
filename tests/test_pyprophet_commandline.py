import os
import subprocess
import shutil
import sys

import pytest

__here__ = os.path.dirname(os.path.abspath(__file__))


def _record(stdout, regtest):
    for line in stdout.split("\n"):
        if line.startswith("INFO"):
            # remove data which is not constant over runs as time or process id:
            line = line.split(":", 4)[-1]
        if "NEEDED" in line:
            continue
        if "time needed" in line:
            continue
        print >> regtest, line


def _run_cmdline(cmdline):
    stdout = cmdline + "\n"
    try:
        stdout += subprocess.check_output(cmdline, shell=True,
                                          stderr=subprocess.STDOUT)
    except subprocess.CalledProcessError, e:
        print >> sys.stderr, e.output
        raise
    return stdout


def _run_pyprophet_to_learn_model(regtest, tmpdir, dump_result_files=False, with_probatilities=False):
    os.chdir(tmpdir.strpath)
    data_path = os.path.join(__here__, "test_data.txt")
    shutil.copy(data_path, tmpdir.strpath)
    if with_probatilities:
        stdout = _run_cmdline("pyprophet test_data.txt --is_test --compute.probabilities")
    else:
        stdout = _run_cmdline("pyprophet test_data.txt --is_test")
    for f in ["test_data_summary_stat.csv", "test_data_full_stat.csv", "test_data_report.pdf",
              "test_data_scorer.bin",
              "test_data_cutoffs.txt",
              "test_data_svalues.txt",
              "test_data_qvalues.txt",
              "test_data_with_dscore.csv",
              "test_data_with_dscore_filtered.csv",
              "test_data_dscores_top_target_peaks.txt",
              "test_data_dscores_top_decoy_peaks.txt",
              ]:
        full_path = os.path.join(tmpdir.strpath, f)
        assert os.path.exists(full_path)

    if not dump_result_files:
        return stdout

    for f in ["test_data_summary_stat.csv", "test_data_full_stat.csv",
              "test_data_cutoffs.txt",
              "test_data_svalues.txt",
              "test_data_qvalues.txt",
              "test_data_dscores_top_target_peaks.txt",
              "test_data_dscores_top_decoy_peaks.txt",
              "test_data_with_dscore.csv",
              "test_data_with_dscore_filtered.csv",
              ]:
        full_path = os.path.join(tmpdir.strpath, f)
        lines = open(full_path, "r").readlines()
        print >> regtest
        print >> regtest, f, "contains", len(lines), "lines"
        print >> regtest
        if len(lines) > 10:
            print >> regtest, "top 5 lines of", f
            for line in lines[:5]:
                if len(line) > 80:
                    line = line[:40] + " ... " + line[-45:]
                print >> regtest, line.rstrip()
            print >> regtest
            print >> regtest, "last 5 lines of", f
            for line in lines[-5:]:
                if len(line) > 80:
                    line = line[:40] + " ... " + line[-45:]
                print >> regtest, line.rstrip()
        else:
            for line in lines:
                if len(line) > 80:
                    line = line[:40] + " ... " + line[-45:]
                print >> regtest, line.rstrip()

    return stdout


def test_0(tmpdir, regtest):
    stdout = _run_pyprophet_to_learn_model(regtest, tmpdir, True, False)
    _record(stdout, regtest)


def test_1(tmpdir, regtest):
    stdout = _run_pyprophet_to_learn_model(regtest, tmpdir, True, True)
    _record(stdout, regtest)


def test_apply_classifier(tmpdir, regtest):

    stdout = _run_pyprophet_to_learn_model(regtest, tmpdir)

    _record(stdout, regtest)

    # collect m score stats
    m_score_stat, = [l for l in stdout.split("\n") if "mean m_score" in l]

    # split away log time etc:
    __, __, interesting_m_score_output = m_score_stat.partition("mean m_score")

    # collect s value stats
    s_value_stat, = [l for l in stdout.split("\n") if "mean s_value" in l]

    # split away log time etc:
    __, __, interesting_s_value_output = s_value_stat.partition("mean s_value")

    stdout = _run_cmdline("pyprophet test_data.txt --apply_scorer=test_data_scorer.bin "
                          "--target.overwrite --is_test")

    _record(stdout, regtest)

    # collect m score stats
    m_score_stat, = [l for l in stdout.split("\n") if "mean m_score" in l]

    # split away log time etc:
    __, __, interesting_m_score_output2 = m_score_stat.partition("mean m_score")

    # collect s value stats
    s_value_stat, = [l for l in stdout.split("\n") if "mean s_value" in l]

    # split away log time etc:
    __, __, interesting_s_value_output2 = s_value_stat.partition("mean s_value")

    # check if output is consistent over runs:
    assert interesting_m_score_output == interesting_m_score_output2
    assert interesting_s_value_output == interesting_s_value_output2


def test_apply_weights(tmpdir, regtest):

    stdout = _run_pyprophet_to_learn_model(regtest, tmpdir)
    _record(stdout, regtest)

    stdout = _run_cmdline("pyprophet test_data.txt --apply_weights=test_data_weights.txt "
                          "--target.overwrite --is_test")

    _record(stdout, regtest)


def test_multiple_input_files(tmpdir, regtest):
    os.chdir(tmpdir.strpath)

    data_path = os.path.join(__here__, "test_data_3.txt")
    shutil.copy(data_path, tmpdir.strpath)
    data_path = os.path.join(__here__, "test_data_2.txt")
    shutil.copy(data_path, tmpdir.strpath)

    stdout = _run_cmdline("pyprophet test_data_2.txt test_data_3.txt --is_test --target.overwrite")
    _record(stdout, regtest)


def test_multiple_input_files_with_merge(tmpdir, regtest):
    os.chdir(tmpdir.strpath)

    data_path = os.path.join(__here__, "test_data_3.txt")
    shutil.copy(data_path, tmpdir.strpath)
    data_path = os.path.join(__here__, "test_data_2.txt")
    shutil.copy(data_path, tmpdir.strpath)

    stdout = _run_cmdline("pyprophet test_data_2.txt test_data_3.txt --is_test --target.overwrite "
                          "--multiple_files.merge_results --compute.probabilities")

    _record(stdout, regtest)

