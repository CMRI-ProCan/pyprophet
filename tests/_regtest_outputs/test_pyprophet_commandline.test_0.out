
test_data_summary_stat.csv contains 10 lines

qvalue	svalue	TP	FP	TN	FN	FDR	sens	cutoff
0.0	0.0216606498192	9.0	0.0	16.709844559 ... 5145469717e-11	0.0243052640416	7.13743586116
0.01	1.0	374.0	3.0	13.7098445596	-3.7098 ... 4455959	0.00683275103368	1.0	-0.291337363603
0.02	1.0	371.0	8.0	8.70984455959	-0.7098 ... 44559585	0.0217885703034	1.0	-0.347789761479
0.05	1.0	373.0	14.0	2.70984455959	-2.709 ... 0984455959	0.037349142127	1.0	-1.33578698687
0.1
0.2
0.3
0.4
0.5

test_data_full_stat.csv contains 52 lines

top 5 lines of test_data_full_stat.csv
qvalue	svalue	TP	FP	TN	FN	FDR	sens	cutoff
0.037349142127	1.0	373.0	14.0	2.70984455 ... 0984455959	0.037349142127	1.0	-1.77522692899
0.037349142127	1.0	373.0	14.0	2.70984455 ... 0984455959	0.037349142127	1.0	-1.58187335446
0.037349142127	1.0	373.0	14.0	2.70984455 ... 0984455959	0.037349142127	1.0	-1.38851977993
0.0368507235919	1.0	371.0	14.0	2.7098445 ... 9844559585	0.0368507235919	1.0	-1.1951662054

last 5 lines of test_data_full_stat.csv
2.17224386626e-11	0.0270758122738	11.0	0 ... 7870166352e-11	0.0297064338287	7.11903749942
1.45145469717e-11	0.0108303249095	4.0	0. ... 7689497714e-11	0.0108023395741	7.31239107395
1.45145469717e-11	0.00541516245477	1.0	0 ... 692913974e-11	0.00270058489352	7.50574464848
1.45145469717e-11	0.00541516245477	1.0	0 ... 692913974e-11	0.00270058489352	7.69909822302
1.45145469717e-11	0.00541516245477	1.0	0 ... 692913974e-11	0.00270058489352	7.89245179755

test_data_cutoffs.txt contains 1 lines

-1.775227e+00 -1.581873e+00 -1.388520e+0 ... 91e+00 7.505745e+00 7.699098e+00 7.892452e+00

test_data_svalues.txt contains 1 lines

1.000000e+00 1.000000e+00 1.000000e+00 1 ... 32e-02 5.415162e-03 5.415162e-03 5.415162e-03

test_data_qvalues.txt contains 1 lines

3.734914e-02 3.734914e-02 3.734914e-02 3 ... 55e-11 1.451455e-11 1.451455e-11 1.451455e-11

test_data_dscores_top_target_peaks.txt contains 1 lines

5.549732e+00 4.755534e+00 4.345843e+00 5 ... 62e+00 6.077132e+00 6.864067e+00 5.938154e+00

test_data_dscores_top_decoy_peaks.txt contains 1 lines

7.452936e-01 2.042587e+00 -3.548498e-01  ... e+00 -3.708231e-01 -6.893719e-02 1.098074e+00

test_data_with_dscore.csv contains 9166 lines

top 5 lines of test_data_with_dscore.csv
transition_group_id	run_id	decoy	main_va ... series_score	d_score	m_score	peak_group_rank
459_run0	0	0	5.27893963303	5	0.993098149 ... 79341002	6	5.54973193397	8.83676159036e-09	1
459_run0	0	0	-0.13974375122	0	0.95392456 ... 6396888252	0	-1.59305365459	0.037349142127	3
459_run0	0	0	-0.77504285962	0	0.66666624 ... 9915833201	0	-7.43544856063	0.037349142127	7
459_run0	0	0	-1.05454511848	0	0.84734030 ... 0811834697	0	-4.75434466144	0.037349142127	6

last 5 lines of test_data_with_dscore.csv
DECOY_592_run0	0	1	-1.84945466301	0	0.63 ... 880632506	0	-3.65461482563	0.037349142127	11
DECOY_592_run0	0	1	-2.10957129817	0	0.90 ... 0335674837	0	-1.52916887883	0.037349142127	4
DECOY_592_run0	0	1	-2.31560973872	0	0.90 ... 766613662	0	-5.10635036746	0.037349142127	16
DECOY_592_run0	0	1	-2.53264045217	0	0.98 ... 796452491	0	-4.08469665226	0.037349142127	14
DECOY_592_run0	0	1	-3.44062881815	0	-1.0 ... 0	0.0	0.0	0	-3.46694393625	0.037349142127	10

test_data_with_dscore_filtered.csv contains 9166 lines

top 5 lines of test_data_with_dscore_filtered.csv
transition_group_id	run_id	decoy	main_va ... series_score	d_score	m_score	peak_group_rank
459_run0	0	0	5.27893963303	5	0.993098149 ... 79341002	6	5.54973193397	8.83676159036e-09	1
459_run0	0	0	-0.13974375122	0	0.95392456 ... 6396888252	0	-1.59305365459	0.037349142127	3
459_run0	0	0	-0.77504285962	0	0.66666624 ... 9915833201	0	-7.43544856063	0.037349142127	7
459_run0	0	0	-1.05454511848	0	0.84734030 ... 0811834697	0	-4.75434466144	0.037349142127	6

last 5 lines of test_data_with_dscore_filtered.csv
DECOY_592_run0	0	1	-1.84945466301	0	0.63 ... 880632506	0	-3.65461482563	0.037349142127	11
DECOY_592_run0	0	1	-2.10957129817	0	0.90 ... 0335674837	0	-1.52916887883	0.037349142127	4
DECOY_592_run0	0	1	-2.31560973872	0	0.90 ... 766613662	0	-5.10635036746	0.037349142127	16
DECOY_592_run0	0	1	-2.53264045217	0	0.98 ... 796452491	0	-4.08469665226	0.037349142127	14
DECOY_592_run0	0	1	-3.44062881815	0	-1.0 ... 0	0.0	0.0	0	-3.46694393625	0.037349142127	10
pyprophet test_data.txt --is_test
 config settings:
     apply_scorer: None
     apply_weights: None
     compute.probabilities: False
     d_score.cutoff: -1000.0
     delim.in:
     delim.out:
     export.mayu: False
     final_statistics.fdr_all_pg: False
     final_statistics.lambda: 0.4
     ignore.invalid_score_columns: 0
     is_test: 1
     multiple_files.merge_results: 0
     num_processes: 1
     semi_supervised_learner.initial_fdr: 0.15
     semi_supervised_learner.initial_lambda: 0.4
     semi_supervised_learner.iteration_fdr: 0.02
     semi_supervised_learner.iteration_lambda: 0.4
     semi_supervised_learner.num_iter: 5
     target.dir: None
     target.overwrite: 0
     target.prefix: None
     xeval.fraction: 0.5
     xeval.num_iter: 5
 process test_data.txt
 learn and apply classifier from input data
 data set contains 387 decoy and 387 target transition groups
 summary input file:
    9165 lines
    774 transition groups
    17 scores including main score
 learn and apply scorer
 start 5 cross evals using 1 processes
 start learn_randomized
 end learn_randomized
 start learn_randomized
 end learn_randomized
 start learn_randomized
 end learn_randomized
 start learn_randomized
 end learn_randomized
 start learn_randomized
 end learn_randomized
 finished cross evals

 data set contains 387 decoy and 387 target transition groups
 mean m_score = 3.179270e-02, std_dev m_score = 1.183096e-02
 mean s_value = 9.785905e-01, std_dev s_value = 1.145276e-01
 calculated scoring and statistics
 processing input data finished

used parameters:

    apply_scorer                               : None
    apply_weights                              : None
    compute.probabilities                      : False
    d_score.cutoff                             : -1000.0
    delim.in                                   : '\t'
    delim.out                                  : '\t'
    export.mayu                                : False
    final_statistics.fdr_all_pg                : False
    final_statistics.lambda                    : 0.4
    ignore.invalid_score_columns               : 0
    is_test                                    : 1
    multiple_files.merge_results               : 0
    num_processes                              : 1
    semi_supervised_learner.initial_fdr        : 0.15
    semi_supervised_learner.initial_lambda     : 0.4
    semi_supervised_learner.iteration_fdr      : 0.02
    semi_supervised_learner.iteration_lambda   : 0.4
    semi_supervised_learner.num_iter           : 5
    target.dir                                 : None
    target.overwrite                           : 0
    target.prefix                              : None
    xeval.fraction                             : 0.5
    xeval.num_iter                             : 5


==================================================================================================

   qvalue   svalue   TP  FP        TN         FN          FDR     sens   cutoff
0    0.00  0.02166    9   0  16.70984  361.29016  1.45145e-11  0.02431  7.13744
1    0.01  1.00000  374   3  13.70984   -3.70984  6.83275e-03  1.00000 -0.29134
2    0.02  1.00000  371   8   8.70984   -0.70984  2.17886e-02  1.00000 -0.34779
3    0.05  1.00000  373  14   2.70984   -2.70984  3.73491e-02  1.00000 -1.33579
4    0.10      NaN  NaN NaN       NaN        NaN          NaN      NaN      NaN
5    0.20      NaN  NaN NaN       NaN        NaN          NaN      NaN      NaN
6    0.30      NaN  NaN NaN       NaN        NaN          NaN      NaN      NaN
7    0.40      NaN  NaN NaN       NaN        NaN          NaN      NaN      NaN
8    0.50      NaN  NaN NaN       NaN        NaN          NaN      NaN      NaN

==================================================================================================

WRITTEN:  test_data_summary_stat.csv
WRITTEN:  test_data_full_stat.csv
WRITTEN:  test_data_report.pdf
WRITTEN:  test_data_cutoffs.txt
WRITTEN:  test_data_svalues.txt
WRITTEN:  test_data_qvalues.txt
WRITTEN:  test_data_dscores_top_target_peaks.txt
WRITTEN:  test_data_dscores_top_decoy_peaks.txt
WRITTEN:  test_data_with_dscore.csv
WRITTEN:  test_data_with_dscore_filtered.csv
WRITTEN:  test_data_scorer.bin
WRITTEN:  test_data_weights.txt


